### ISSUE TITLE
MD046 assumes block indents are bad if they are a part of lists.

### ISSUE THREAD

**Issue by jamesharris-garmin** [DavidAnson/markdownlint#327]
## Description

MD046 incorectly flags all fenced code blocks as violations even if they are a part of lists.

### Motivating example:

```markdown
1. Here is my list item

    ```bash
    # example code
    ```
2. Here is my next list item.
```
which I expect to render as follows:

---

1. Here is my list item

    ```bash
    # example code
    ```
2. Here is my next list item.

---

The vscode markdownlint extension flags this as a violation of MD046  but if I follow the advice on the [MD046](https://github.com/DavidAnson/markdownlint/blob/v0.20.4/doc/Rules.md#md046) wiki page then I get the following rendering:

### Example following [MD046](https://github.com/DavidAnson/markdownlint/blob/v0.20.4/doc/Rules.md#md046) advice:
```markdown
1. Here is my list item

```bash
# example code
``` <!-- space added here to prevent rendering errors in comment-->

2. Here is my next list item.
```
Rendered result:

---
1. Here is my list item

```bash
# example code
```

2. Here is my next list item.

---

which is not how I want the list to appear in its rendered form.

### Possible Resolutions:

It seems like there are two approaches that could be used to fix this issue:

1. broaden MD046 to allow for these cases when in lists. (both formations are valid so it probably just needs to be disabled when immediately following list items (ordered or unordered)
2. Allow the visual studio code extension to ignore specific violations.

**Comment by DavidAnson** [comment]
Here is your motivating example. It does not seem to trigger a violation of MD046. Could you please clarify the scenario?

https://dlaa.me/markdownlint/#%25m%23%20Issue%20327%0A%0A1.%20Here%20is%20my%20list%20item%0A%0A%20%20%20%20%60%60%60bash%0A%20%20%20%20%23%20example%20code%0A%20%20%20%20%60%60%60%0A2.%20Here%20is%20my%20next%20list%20item.%0A

Regarding your desire to suppress individual violations, that is possible today and documented here: https://github.com/DavidAnson/vscode-markdownlint#suppress

**Comment by DavidAnson** [comment]
Maybe it's helpful to clarify the intent of rule MD046. It is a stylistic rule to enforce the  same technique (indenting or fencing) throughout the document. By default, it assumes the first style it encounters is correct and flags any differences in the rest of the document. In your case, if you have an indented code block earlier in the document, then the fenced code you show above is a violation because it is not an indented code block

**Comment by MichaIng** [comment]
Interesting, we faced the same issue, but when copying our code into the demo editor, the violation is not shown:
Demo: https://dlaa.me/markdownlint/#%25m%23%20Issue%20327%0A%0A1.%20Here%20is%20my%20list%20item%0A%0A%20%20%20%20%60%60%60bash%0A%20%20%20%20%23%20example%20code%0A%20%20%20%20%60%60%60%0A2.%20Here%20is%20my%20next%20list%20item.%0A
Failed check: https://github.com/MichaIng/DietPi-Docs/runs/1272589173?check_suite_focus=true#step:6:5
Code: https://raw.githubusercontent.com/MichaIng/DietPi-Docs/c4a9836a43efc386ff9c6ae1d5f61c6a439ca5d4/docs/release-notes.md

**Comment by DavidAnson** [comment]
@MichaIng Your failing test run reports an issue with `docs/dietpi_tools.md`, but your example and link are for a different file.

Looking at the reported file, we see code fences on lines 7 and 17, so the report of an indented code block on line 27 being inconsistent is correct.

https://github.com/MichaIng/DietPi-Docs/blame/b9ac99eab398f0b850eac7460ea8267f06bf0c4e/docs/dietpi_tools.md

**Comment by MichaIng** [comment]
Ah indeed, sorry for mixing this up. This makes all sense now, the failing code lines are due to [pymdownx.tabbed](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/) Markdown extension for content tabs. I guess adding support for this would require a switch/parameter since it is relatively rarely used compared to nested fences via [pymdownx.superfences](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/). Was this already a topic once or shall I open a request for it?

Related to the original topic I can hence confirm that nested fences via four-space indentation is correctly handled by markdownlint.

**Comment by DavidAnson** [comment]
Thanks, I will close this issue. Regarding tabbed behavior, that syntax seems to be a violation of CommonMark syntax, so it's not clear to me how to do better unless there is a markdown-it parser to support that. You're welcome to open an issue, but FYI that there are complications.

**Comment by MichaIng** [comment]
Many thanks for mentioning `CommonMark`, good to know there is such a specification, only sad that Python-Markdown core + standard library extensions do not fulfil it completely (e.g. fenced code in list items is not possible) so that pymdownx needs to be used for that, while other extensions of pymdownx are not part of CommonMark. Not a real issue but that makes it difficult for newcomers to differentiate between common/specified syntax that is usually interpreted correctly and uncommon syntax where one has to expect failures of code quality/syntax tests 😅.

But all fine, I see that these content tabs a very uncommon and conflict with some header and thematic break syntax. So the effort to implement support is likely too large effort to be reasonable.
